src/clr-angular/utils/popover/adapter/operators.ts

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+import { ClrPopoverPosition } from '../interfaces/popover-position.interface';
+import { ClrAxis } from '../enums/axis.enum';
+import { ClrSide } from '../enums/side.enum';
+import { ClrAlignment } from '../enums/alignment.enum';
+
+export const SIMPLE_TO_SMART_POSITIONS: { [input: string]: ClrPopoverPosition } = {
+  'top-left': {
+    axis: ClrAxis.VERTICAL,
+    side: ClrSide.BEFORE,
+    anchor: ClrAlignment.CENTER,
+    content: ClrAlignment.END,
+  },
+  'top-middle': {
+    axis: ClrAxis.VERTICAL,
+    side: ClrSide.BEFORE,
+    anchor: ClrAlignment.CENTER,
+    content: ClrAlignment.CENTER,
+  },
+  'top-right': {
+    axis: ClrAxis.VERTICAL,
+    side: ClrSide.BEFORE,
+    anchor: ClrAlignment.CENTER,
+    content: ClrAlignment.START,
+  },
+  'right-top': {
+    axis: ClrAxis.HORIZONTAL,
+    side: ClrSide.AFTER,
+    anchor: ClrAlignment.CENTER,
+    content: ClrAlignment.END,
+  },
+  'right-middle': {
+    axis: ClrAxis.HORIZONTAL,
+    side: ClrSide.AFTER,
+    anchor: ClrAlignment.CENTER,
+    content: ClrAlignment.CENTER,
+  },
+  'right-bottom': {
+    axis: ClrAxis.HORIZONTAL,
+    side: ClrSide.AFTER,
+    anchor: ClrAlignment.CENTER,
+    content: ClrAlignment.START,
+  },
+  'bottom-right': {
+    axis: ClrAxis.VERTICAL,
+    side: ClrSide.AFTER,
+    anchor: ClrAlignment.CENTER,
+    content: ClrAlignment.START,
+  },
+  'bottom-middle': {
+    axis: ClrAxis.VERTICAL,
+    side: ClrSide.AFTER,
+    anchor: ClrAlignment.CENTER,
+    content: ClrAlignment.CENTER,
+  },
+  'bottom-left': {
+    axis: ClrAxis.VERTICAL,
+    side: ClrSide.AFTER,
+    anchor: ClrAlignment.CENTER,
+    content: ClrAlignment.END,
+  },
+  'left-bottom': {
+    axis: ClrAxis.HORIZONTAL,
+    side: ClrSide.BEFORE,
+    anchor: ClrAlignment.CENTER,
+    content: ClrAlignment.START,
+  },
+  'left-middle': {
+    axis: ClrAxis.HORIZONTAL,
+    side: ClrSide.BEFORE,
+    anchor: ClrAlignment.CENTER,
+    content: ClrAlignment.CENTER,
+  },
+  'left-top': {
+    axis: ClrAxis.HORIZONTAL,
+    side: ClrSide.BEFORE,
+    anchor: ClrAlignment.CENTER,
+    content: ClrAlignment.END,
+  },
+};
+
+export function posiblePositionsAsStrings() {
+  return Object.keys(SIMPLE_TO_SMART_POSITIONS);
+}
+
+export function defaultPosition() {
+  return SIMPLE_TO_SMART_POSITIONS['bottom-right'];
+}
+export function transposePositionString(position: ClrPopoverPosition): string {
```

Comment:
I'm not sure why this is a separate file instead of part of the position-operators.ts. These functions all operate on ClrPositions (or the string representation of a position).

Reply:
they could go there this is not a problem, 
But I could argue that the naming of the method is quite clear for example:

```js
transposePositionString()
transposeStringPosition()
```
They look too similar and sound similar but are getting different arguments.

Could go with something like that:
```js
transposePositionObject()
transposePositionString()
```
Or even use the word `find` cause that is what we do, going over a list of elements to find a match.

```js
findPositionObject('bottom-left');
findPositionString(ClrPopoverPosition);
```

Reply:
That's ok as well :) just didn't like the `PositionString`, `StringPosition` I got the idea just was a bit cryptic of don't know the code.

Reply:
i like `find`. Its a little clearer whats going on vs `transpose`.

Reply:
Are we all agreeing on `findPositionObject` and `findPositionString` ?